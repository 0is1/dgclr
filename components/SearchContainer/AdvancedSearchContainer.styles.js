import styled from 'styled-components';

const Styles = {
  ToggleButtonWrapper: styled.div`
    padding: 0.25rem;
    position: absolute;
    right: 0;
    top: 0;
    width: 30px;
    &:hover {
      cursor: pointer;
    }
  `,
};

export default Styles;
