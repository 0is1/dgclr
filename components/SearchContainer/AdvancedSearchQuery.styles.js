import styled from 'styled-components';
import colors from 'components/colors';

const Styles = {
  NoResults: styled.h2`
    color: ${colors.red};
    padding: 0.25rem;
    text-align: center;
  `,
};

export default Styles;
