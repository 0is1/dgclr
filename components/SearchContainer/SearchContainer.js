// @flow
import React, { PureComponent } from 'react';
import { debounce } from 'lodash';
import Container from 'components/Container';
import Input from 'components/Input';
import SearchQuery from 'components/SearchContainer/SearchQuery';
import Styles from 'components/SearchContainer/SearchContainer.styles';

const { Wrapper } = Styles;

type Props = {
  client: {},
};
type State = {
  inputValue: string,
  query: string,
};

class SearchContainer extends PureComponent<Props, State> {
  state = {
    inputValue: '',
    query: '',
  };

  changeQueryValue = debounce((query) => {
    this.setState({ query });
  }, 300);

  onChange = (query: string) => {
    console.log('value is: ', query);
    this.setState({ inputValue: query });
    this.changeQueryValue(query);
  };

  render() {
    const { query, inputValue } = this.state;
    return (
      <Container>
        <Wrapper>
          <h1>Etsi ratoja:</h1>
          <Input
            placeholder="Kaupungin tai radan nimi"
            value={inputValue}
            onChange={this.onChange}
          />
          <SearchQuery query={query} />
        </Wrapper>
      </Container>
    );
  }
}

export default SearchContainer;
