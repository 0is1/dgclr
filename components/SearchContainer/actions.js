// @flow
import type { Course, FilterData, MapFilterData } from 'lib/types';

export const actionTypes = {
  SET_COURSES: 'DGCLR_SET_COURSES',
  SET_SEARCH_QUERY: 'DGCLR_SET_SEARCH_QUERY',
  TOGGLE_ADVANCED_SEARCH: 'DGCLR_TOGGLE_ADVANCED_SEARCH',
  SET_ADVANCED_SEARCH_QUERY: 'DGCLR_SET_ADVANCED_SEARCH_QUERY',
  SET_ADVANCED_FILTER_TYPE: 'DGCLR_SET_ADVANCED_FILTER_TYPE',
  SET_CURRENT_ADVANCED_FILTER: 'DGCLR_SET_CURRENT_ADVANCED_FILTER',
};

/**
 * Set courses action
 * @param {Array} courses
 * @return {Object} action of SET_COURSES
 */
export const setCourses = (courses: Array<Course> = []) => ({
  type: actionTypes.SET_COURSES,
  courses,
});

/**
 * Set search query action
 * @param {Array} courses
 * @param {String} query
 * @return {Object} action of SET_SEARCH_QUERY
 */
export const setSearchQuery = (courses: Array<Course> = [], query: string) => ({
  type: actionTypes.SET_SEARCH_QUERY,
  courses,
  query,
});

/**
 * Toggle advanced search open
 * @param {boolean} open
 * @return {Object} action of TOGGLE_ADVANCED_SEARCH
 */
export const toggleAdvancedSearch = (open: boolean = false) => ({
  type: actionTypes.TOGGLE_ADVANCED_SEARCH,
  open,
});

/**
 * Set filter data to store
 * @param {string} filterName
 * @param {Array} data
 * @return {Object} action of SET_ADVANCED_BASKET_TYPE_FILTER
 */
export const setFilter = (filterName: string = '', data: FilterData | MapFilterData) => ({
  type: actionTypes.SET_ADVANCED_FILTER_TYPE,
  filterName,
  data,
});

/**
 * Set advanced search query action
 * @param {Array} courses
 * @param {String} query
 * @return {Object} action of SET_ADVANCED_SEARCH_QUERY
 */
export const setAdvancedSearchQuery = (courses: Array<Course> = [], query: string) => ({
  type: actionTypes.SET_ADVANCED_SEARCH_QUERY,
  courses,
  query,
});

/**
 * Set current advanced search filters
 * @param {String} filter stringified JSON
 * @return {Object} action of SET_CURRENT_ADVANCED_FILTER
 */
export const setCurrentAdvancedSearchFilter = (filter: string = '') => ({
  type: actionTypes.SET_CURRENT_ADVANCED_FILTER,
  filter,
});
