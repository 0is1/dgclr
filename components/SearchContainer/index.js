// @flow
import SearchContainerComponent from './SearchContainer';

export default SearchContainerComponent;
