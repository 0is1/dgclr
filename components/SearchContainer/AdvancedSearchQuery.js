// @flow
import React, { Component } from 'react';
import gql from 'graphql-tag';
import { graphql } from 'react-apollo';
import Link from 'next/link';
import { connect } from 'react-redux';
import { differenceBy, size } from 'lodash';
import {
  setCourses as setCoursesFunc,
  setAdvancedSearchQuery,
} from 'components/SearchContainer/actions';
import {
  latestAdvancedQuery as latestQueryFunc,
  queryResultsFromState as queryResultsFromStateFunc,
} from 'components/SearchContainer/selectors';
import { getRandomKey, uniqueLayoutRatings } from 'helpers/utils';
import LayoutRatingBadges from 'components/Layout/Badges';
import { ClipLoader } from 'components/Spinners';
import Styles from 'components/SearchContainer/SearchContainer.styles';
import BaseStyles from 'components/Container/Container.styles';
import type { GraphQLData, State } from 'lib/types';

const { UL, LI } = BaseStyles;

const { SearchResultItem } = Styles;

type Props = {
  filter: {},
  queryResults: [],
  data: GraphQLData,
  latestQuery: string,
  setCourses: Function,
  setSearchQuery: Function,
};

class AdvancedSearchQuery extends Component<Props> {
  componentDidUpdate(prevProps, prevState, snapshot) {
    if (snapshot !== null) {
      const { data = {} } = this.props;
      const { courses = [] } = data;
      if (courses) {
        this.setCourses(courses);
      }
    }
  }

  getSnapshotBeforeUpdate(prevProps) {
    const { data = {}, latestQuery, filter } = this.props;
    const { courses = [] } = data;
    const prevCourseData = prevProps.data;
    const prevCourseArr = prevCourseData && prevProps.data.courses;
    const prevCourses = prevCourseArr && prevCourseArr.length ? prevCourseArr : [];
    const stringifyFilter = JSON.stringify(filter);
    // console.log('courses: ', courses);
    // console.log('prevCourses: ', prevCourses);
    if (
      (courses && courses.length && differenceBy(courses, prevCourses, '_id').length > 0)
      || (!courses.length && stringifyFilter !== latestQuery)
      || (courses.length !== prevCourses.length && stringifyFilter !== latestQuery)
    ) {
      return true;
    }
    return null;
  }

  setCourses = (courses) => {
    const { setCourses, setSearchQuery, filter } = this.props;
    setCourses(courses);
    setSearchQuery(courses, JSON.stringify(filter));
  };

  render() {
    const { filter, queryResults, data = {} } = this.props;
    if (size(filter) < 1 && !queryResults.length) return null;
    if (data && data.loading) {
      return <ClipLoader />;
    }
    const { courses } = data;
    const courseData = queryResults.length ? queryResults : courses;
    let results = <li>Ei hakutuloksia!</li>;
    if (courseData && courseData.length) {
      results = courseData.map((course) => {
        const ratings = uniqueLayoutRatings(course.layouts);
        return (
          <LI key={getRandomKey()}>
            <Link as={`/${course.slug}`} href={`/course?slug=${course.slug}`}>
              <SearchResultItem>
                {course.name}
                <LayoutRatingBadges tiny ratings={ratings} />
              </SearchResultItem>
            </Link>
          </LI>
        );
      });
    }
    return <UL>{results}</UL>;
  }
}

const mapStateToProps = (state: State) => ({
  latestQuery: latestQueryFunc(state),
  queryResults: queryResultsFromStateFunc(state, latestQueryFunc(state)),
});

const mapDispatchToProps = dispatch => ({
  setCourses: courses => dispatch(setCoursesFunc(courses)),
  setSearchQuery: (courses, query) => dispatch(setAdvancedSearchQuery(courses, query)),
});

const SEARCH_COURSES = gql`
  query AdvancedCoursesQuery($filter: CourseQueryFilterInput!) {
    courses(filter: $filter) {
      _id
      name
      description
      slug
      courseInfo {
        basketType
        teeType
        infoSignType
        surfaceShapeTypes
        courseTypes
        mapUrl
        founded
        maintenanceCycle
        rangeMaster
        courseDesigner
        fee {
          amount
          currency
          value
        }
      }
      locationInfo {
        address
        city
        zip
        location {
          coordinates
        }
      }
      layouts {
        name
        rating
        holes {
          par
          length {
            foot
            meter
          }
        }
        holeCount
        totalPar
      }
    }
  }
`;

// The `graphql` wrapper executes a GraphQL query and makes the results
// available on the `data` prop of the wrapped component (PostList)
const ComponentWithMutation = graphql(SEARCH_COURSES, {
  options: ({ filter }) => ({
    variables: {
      filter,
    },
  }),
  props: ({ data }) => ({
    data,
  }),
  skip: ({ filter }) => !filter || size(filter) < 1,
})(AdvancedSearchQuery);

export default connect(
  mapStateToProps,
  mapDispatchToProps,
)(ComponentWithMutation);
