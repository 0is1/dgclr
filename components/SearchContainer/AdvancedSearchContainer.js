// @flow
import React, { Component } from 'react';
import { connect } from 'react-redux';
import { Box } from 'rebass';
import { GoSettings } from 'react-icons/go';
import { toggleAdvancedSearch as toggleAdvancedSearchFunc } from 'components/SearchContainer/actions';
import { isAdvancedSearchOpen } from 'components/SearchContainer/selectors';
import AdvancedSearchInputs from 'components/SearchContainer/AdvancedSearchInputs';
import Styles from 'components/SearchContainer/AdvancedSearchContainer.styles';
import colors from 'components/colors';

const { ToggleButtonWrapper } = Styles;

type Props = { open: boolean, toggleAdvancedSearch: Function };

class AdvancedSearchContainer extends Component<Props> {
  toggleOpen = () => {
    const { open, toggleAdvancedSearch } = this.props;
    toggleAdvancedSearch(!open);
  };

  handleKeyPress = () => {};

  render() {
    const { open } = this.props;
    const inputs = open ? <AdvancedSearchInputs /> : null;
    const settingsColor = open ? colors.info : colors.text;
    return (
      <Box style={{ position: 'relative' }}>
        {inputs}
        <ToggleButtonWrapper
          onClick={this.toggleOpen}
          onKeyPress={this.handleKeyPress}
          role="button"
          tabIndex="0"
        >
          <GoSettings size="1.5rem" color={settingsColor} />
        </ToggleButtonWrapper>
      </Box>
    );
  }
}

const mapStateToProps = state => ({
  open: isAdvancedSearchOpen(state),
});

const mapDispatchToProps = dispatch => ({
  toggleAdvancedSearch: (open = false) => dispatch(toggleAdvancedSearchFunc(open)),
});

export default connect(
  mapStateToProps,
  mapDispatchToProps,
)(AdvancedSearchContainer);
