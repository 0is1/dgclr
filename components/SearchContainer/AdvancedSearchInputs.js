// @flow
import React, { Component } from 'react';
import { connect } from 'react-redux';
import { Box, Flex, Text } from 'rebass';
import update from 'updeep';
import { omit, size } from 'lodash';
import { getCurrentAdvancedFilter } from 'components/SearchContainer/selectors';
import { setCurrentAdvancedSearchFilter } from 'components/SearchContainer/actions';
import RatingSelect from 'components/Select/RatingSelect';
import BasketTypeSelect from 'components/Select/BasketTypeSelect';
import TeeTypeSelect from 'components/Select/TeeTypeSelect';
import SurfaceTypeSelect from 'components/Select/SurfaceTypeSelect';
import AdvancedSearchMap from 'components/Map/AdvancedSearchMap';
import colors from 'components/colors';
import type { CoordinatesObject, State } from 'lib/types';
import { convertMetersToKilometers } from 'helpers/utils';
import {
  ADVANCED_RATING,
  ADVANCED_NEARBY,
  ADVANCED_TEE_TYPE,
  ADVANCED_BASKET_TYPE,
  ADVANCED_COURSE_INFO,
  ADVANCED_SURFACE_SHAPE_TYPES,
} from 'lib/constants';
import RebassComponents from 'components/RebassComponents';

const { Divider } = RebassComponents;
type Props = { filter: string, setFilter: Function };

class AdvancedSearchInputs extends Component<Props> {
  getParsedFilter = () => {
    const { filter } = this.props;
    return !filter.length ? {} : JSON.parse(filter);
  };

  setFilterData = (newFilter) => {
    const { setFilter } = this.props;
    // console.log('newFilter: ', this.cleanCourseInfo(newFilter));
    setFilter(JSON.stringify(this.cleanCourseInfo(newFilter)));
  };

  cleanCourseInfo = (newFilter) => {
    if (newFilter.courseInfo && size(newFilter.courseInfo) === 0) return omit(newFilter, [ADVANCED_COURSE_INFO]);
    return newFilter;
  };

  onRatingChange = (rating: []) => {
    // console.log('onRatingChange: ', rating);
    const filter = this.getParsedFilter();
    // console.log('onRatingChange filter: ', filterData);
    const newFilter = rating.length > 0 ? update({ rating }, filter) : omit(filter, [ADVANCED_RATING]);
    this.setFilterData(newFilter);
  };

  onBasketTypeChange = (basketType: string) => {
    const filter = this.getParsedFilter();
    // console.log('onBasketTypeChange filter: ', filter);
    const newFilter = basketType.length > 0
      ? update({ courseInfo: { basketType } }, filter)
      : { ...filter, courseInfo: omit(filter.courseInfo, [ADVANCED_BASKET_TYPE]) };
    this.setFilterData(newFilter);
  };

  onTeeTypeChange = (teeType: string) => {
    const filter = this.getParsedFilter();
    // console.log('onTeeTypeChange filter: ', filter);
    const newFilter = teeType.length > 0
      ? update({ courseInfo: { teeType } }, filter)
      : { ...filter, courseInfo: omit(filter.courseInfo, [ADVANCED_TEE_TYPE]) };
    this.setFilterData(newFilter);
  };

  onSurfaceTypeChange = (surfaceShapeTypes: string) => {
    const filter = this.getParsedFilter();
    // console.log('surfaceType filter: ', filter);
    const newFilter = surfaceShapeTypes.length > 0
      ? update({ courseInfo: { surfaceShapeTypes } }, filter)
      : { ...filter, courseInfo: omit(filter.courseInfo, [ADVANCED_SURFACE_SHAPE_TYPES]) };
    this.setFilterData(newFilter);
  };

  onMapSearchChange = (data: { coordinates: CoordinatesObject, radius: number }) => {
    const filter = this.getParsedFilter();
    const nearby = data.coordinates
      && data.radius && {
      maxDistance: convertMetersToKilometers(parseInt(data.radius, 10)),
      coordinates: [data.coordinates.lat, data.coordinates.lng],
    };
    const newFilter = nearby ? update({ nearby }, filter) : { ...filter, nearby: omit(filter, [ADVANCED_NEARBY]) };
    this.setFilterData(newFilter);
  };

  render() {
    return (
      <Box>
        <Text fontWeight="700" my={2} width="100%">
          Edistynyt haku (eritt√§in varhaisessa kehitysvaiheessa)
        </Text>
        <Flex flexWrap="wrap">
          <Box pr={[0, 0, '.5rem', '.5rem']} mb=".75rem" width={[1, 1, 1 / 2, 1 / 2]}>
            <RatingSelect onChange={this.onRatingChange} />
          </Box>
          <Box pl={[0, 0, '.5rem', '.5rem']} mb=".75rem" width={[1, 1, 1 / 2, 1 / 2]}>
            <BasketTypeSelect onChange={this.onBasketTypeChange} />
          </Box>
          <Box pr={[0, 0, '.5rem', '.5rem']} mb=".75rem" width={[1, 1, 1 / 2, 1 / 2]}>
            <TeeTypeSelect onChange={this.onTeeTypeChange} />
          </Box>
          <Box pl={[0, 0, '.5rem', '.5rem']} mb=".75rem" width={[1, 1, 1 / 2, 1 / 2]}>
            <SurfaceTypeSelect onChange={this.onSurfaceTypeChange} />
          </Box>
          <Box pl={[0, 0, '.5rem', '.5rem']} mb=".75rem" width={[1]}>
            <AdvancedSearchMap onChange={this.onMapSearchChange} />
          </Box>
        </Flex>
        <Divider w={1} borderColor={colors.info} />
      </Box>
    );
  }
}

const mapStateToProps = (state: State) => ({
  filter: getCurrentAdvancedFilter(state),
});
const mapDispatchToProps = dispatch => ({
  setFilter: filter => dispatch(setCurrentAdvancedSearchFilter(filter)),
});

export default connect(
  mapStateToProps,
  mapDispatchToProps,
)(AdvancedSearchInputs);
