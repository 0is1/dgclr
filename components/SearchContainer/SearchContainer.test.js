import React from 'react';
import { shallow } from 'enzyme';
import { SearchContainer } from './SearchContainer';

const mockProps = (props = {}) => ({
  t: jest.fn(value => value),
  latestQuery: '',
  ...props,
});

describe('<SearchContainer />', () => {
  describe('snapshots', () => {
    it('renders with default props', () => {
      const props = mockProps();
      const subject = shallow(<SearchContainer {...props} />);
      expect(subject).toMatchSnapshot();
    });
    it('renders with latestQuery prop', () => {
      const props = mockProps({ latestQuery: 'espoo' });
      const subject = shallow(<SearchContainer {...props} />);
      expect(subject).toMatchSnapshot();
    });
  });
});
