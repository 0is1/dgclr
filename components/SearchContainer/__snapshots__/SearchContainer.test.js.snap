// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SearchContainer /> snapshots renders with default props 1`] = `
<SearchContainerstyles__Wrapper>
  <Box
    m="1rem auto"
    px="2rem"
    style={
      Object {
        "position": "relative",
      }
    }
    width={
      Array [
        1,
        1,
        1,
        0.7,
      ]
    }
  >
    <Leadstyles
      my={2}
    >
      search:title
    </Leadstyles>
    <InputComponent
      focusOnMount={true}
      onChange={[Function]}
      options={
        Object {
          "type": "text",
        }
      }
      placeholder="search:input-placeholder"
      value=""
    />
    <Connect(withI18nextTranslation(Apollo(SearchQuery)))
      query=""
    />
  </Box>
</SearchContainerstyles__Wrapper>
`;

exports[`<SearchContainer /> snapshots renders with latestQuery prop 1`] = `
<SearchContainerstyles__Wrapper>
  <Box
    m="1rem auto"
    px="2rem"
    style={
      Object {
        "position": "relative",
      }
    }
    width={
      Array [
        1,
        1,
        1,
        0.7,
      ]
    }
  >
    <Leadstyles
      my={2}
    >
      search:title
    </Leadstyles>
    <InputComponent
      focusOnMount={true}
      onChange={[Function]}
      options={
        Object {
          "type": "text",
        }
      }
      placeholder="search:input-placeholder"
      value="espoo"
    />
    <Connect(withI18nextTranslation(Apollo(SearchQuery)))
      query=""
    />
  </Box>
</SearchContainerstyles__Wrapper>
`;
